### PR TITLE
new_owner: a ddl puller wrapper for owner

### PR DESCRIPTION
**Pull request by leoppro**
### What problem does this PR solve? 
a ddl puller wrapper for the new owner, the wrapper starts a puller and listen to the DDL range, add the received DDL into an internal queue. FrontDDL and PopFrontDDL can be used to get and pop the first DDL job in the queue

### Check List 

Tests 

 - Unit test

### Release note

- No release note

### CODE REVIEW

File: cdc/owner/ddl_puller.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+	cdcContext "github.com/pingcap/ticdc/pkg/context"
+	"github.com/pingcap/ticdc/pkg/filter"
+	"github.com/pingcap/ticdc/pkg/regionspan"
+	"github.com/pingcap/ticdc/pkg/util"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+// DDLPuller is a wrapper of Puller for owner
+// DDLPuller start a puller and listen the DDL range, add the received DDL into a internal queue
```
Comment (leoppro) [inline cdc/owner/ddl_puller.go:36]:
```suggestion
// DDLPuller is a wrapper of the Puller interface for the owner
// DDLPuller start a puller and listen to the DDL range, add the received DDL into an internal queue
```

File: cdc/owner/ddl_puller.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+func newDDLPuller(ctx cdcContext.Context, startTs uint64) (DDLPuller, error) {
+	pdCli := ctx.GlobalVars().PDClient
+	conf := config.GetGlobalServerConfig()
+	kvStorage := ctx.GlobalVars().KVStorage
+	f, err := filter.NewFilter(ctx.ChangefeedVars().Info.Config)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var plr puller.Puller
+	// kvStorage can be nil only in test
```
Comment (leoppro) [inline cdc/owner/ddl_puller.go:67]:
```suggestion
	// kvStorage can be nil only in the test
```

File: cdc/owner/ddl_puller.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	puller puller.Puller
+	filter *filter.Filter
+
+	mu             sync.Mutex
+	resolvedTS     uint64
+	pendingDDLJobs []*timodel.Job
+	cancel         context.CancelFunc
+}
+
+func newDDLPuller(ctx cdcContext.Context, startTs uint64) (DDLPuller, error) {
```
Comment (ben1009) [inline cdc/owner/ddl_puller.go:58]: do not find usage in this repo, not enable yet ?

File: cdc/owner/ddl_puller.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+func (h *ddlPullerImpl) Run(ctx cdcContext.Context) error {
+	ctx, cancel := cdcContext.WithCancel(ctx)
+	h.cancel = cancel
+	log.Debug("DDL puller started")
+	stdCtx := util.PutTableInfoInCtx(ctx, -1, "DDL_PULLER")
+	errg, stdCtx := errgroup.WithContext(stdCtx)
+	ctx = cdcContext.WithStd(ctx, stdCtx)
+
+	errg.Go(func() error {
+		return h.puller.Run(ctx)
```
Comment (ben1009) [inline cdc/owner/ddl_puller.go:91]:
```
if kvStorage != nil {
		plr = puller.NewPuller(ctx, pdCli, conf.Security, kvStorage, startTs,
			[]regionspan.Span{regionspan.GetDDLSpan(), regionspan.GetAddIndexDDLSpan()},
			nil, false)
	}

	return &ddlPullerImpl{
		puller:     plr,
		resolvedTS: startTs - 1,
		filter:     f,
		cancel:     func() {},
	}, nil
```
if kvStorage == nil , puller will be nil ?

File: pkg/context/context.go [hunk 5]
```diff
@@ -173,13 +176,16 @@ func (ctx *throwContext) Throw(err error) {
 func NewBackendContext4Test(withChangefeedVars bool) Context {
```
Comment (ben1009) [inline pkg/context/context.go:176]: maybe move test to _test file instead

File: cdc/owner/ddl_puller.go (hunk 3 above)
Comment (leoppro) [reply cdc/owner/ddl_puller.go:58]: yes

File: cdc/owner/ddl_puller.go (hunk 4 above)
Comment (leoppro) [reply cdc/owner/ddl_puller.go:91]: kvStorage can be nil only in test

File: pkg/context/context.go (hunk 5 above)
Comment (leoppro) [reply pkg/context/context.go:176]: we use the `NewBackendContext4Test` function in another package, but a public function in `_test` file can't be import

File: cdc/owner/ddl_puller.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+	return &ddlPullerImpl{
+		puller:     plr,
+		resolvedTS: startTs - 1,
+		filter:     f,
+		cancel:     func() {},
+	}, nil
+}
+
+func (h *ddlPullerImpl) Run(ctx cdcContext.Context) error {
+	ctx, cancel := cdcContext.WithCancel(ctx)
```
[the remark this answers is not in the document]
Comment (leoppro) [reply cdc/owner/ddl_puller.go:83]: I don't think so, the parent ctx should be cover by child ctx here, so we just name them as the same

File: cdc/owner/ddl_puller.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	// kvStorage can be nil only in the test
+	if kvStorage != nil {
+		plr = puller.NewPuller(ctx, pdCli, conf.Security, kvStorage, startTs,
+			[]regionspan.Span{regionspan.GetDDLSpan(), regionspan.GetAddIndexDDLSpan()},
+			nil, false)
+	}
+
+	return &ddlPullerImpl{
+		puller:     plr,
+		resolvedTS: startTs - 1,
```
[the remark this answers is not in the document]
Comment (leoppro) [reply cdc/owner/ddl_puller.go:78]: add comments for it

File: cdc/owner/ddl_puller.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+const ddlPullerName = "DDL_PULLER"
+
+func (h *ddlPullerImpl) Run(ctx cdcContext.Context) error {
+	ctx, cancel := cdcContext.WithCancel(ctx)
+	h.cancel = cancel
+	log.Debug("DDL puller started")
+	stdCtx := util.PutTableInfoInCtx(ctx, -1, ddlPullerName)
+	errg, stdCtx := errgroup.WithContext(stdCtx)
+	ctx = cdcContext.WithStd(ctx, stdCtx)
```
Comment (amyangfei) [inline cdc/owner/ddl_puller.go:92]: The context used in the following `h.puller.Run()` and `puller.SortOutput()` are still standard context, could we use `cdc context` until we need it.

File: cdc/owner/ddl_puller.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+		if job == nil {
+			return nil
+		}
+		if h.filter.ShouldDiscardDDL(job.Type) {
+			log.Info("discard the ddl job", zap.Int64("jobID", job.ID), zap.String("query", job.Query))
+			return nil
+		}
+		if job.BinlogInfo.FinishedTS == lastDDLFinishedTs {
+			return nil
+		}
```
[the remark this answers is not in the document]
Comment (leoppro) [reply cdc/owner/ddl_puller.go:126]: yes, The purpose of L124-L126 is filter the repetitive DDL event

File: cdc/owner/ddl_puller.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+	mu             sync.Mutex
+	resolvedTS     uint64
+	pendingDDLJobs []*timodel.Job
+	cancel         context.CancelFunc
+}
+
+func newDDLPuller(ctx cdcContext.Context, startTs uint64) (DDLPuller, error) {
+	pdCli := ctx.GlobalVars().PDClient
+	conf := config.GetGlobalServerConfig()
+	kvStorage := ctx.GlobalVars().KVStorage
```
Comment (JinLingChristopher) [inline cdc/owner/ddl_puller.go:61]: how about mvoe kvStorage to L67, make related code more compact together.

Comment (leoppro) [reply cdc/owner/ddl_puller.go]: resolved

File: cdc/owner/ddl_puller.go [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+	cdcContext "github.com/pingcap/ticdc/pkg/context"
+	"github.com/pingcap/ticdc/pkg/filter"
+	"github.com/pingcap/ticdc/pkg/regionspan"
+	"github.com/pingcap/ticdc/pkg/util"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+// DDLPuller is a wrapper of the Puller interface for the owner
+// DDLPuller start a puller and listen to the DDL range, add the received DDL into an internal queue
```
Comment (liuzix) [inline cdc/owner/ddl_puller.go:36]:
```suggestion
// DDLPuller starts a puller, listens to the DDL range, adds the received DDLs into an internal queue
```

File: cdc/owner/ddl_puller.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+	// kvStorage can be nil only in the test
+	if kvStorage != nil {
+		plr = puller.NewPuller(ctx, pdCli, conf.Security, kvStorage, startTs,
+			[]regionspan.Span{regionspan.GetDDLSpan(), regionspan.GetAddIndexDDLSpan()},
+			nil, false)
+	}
+
+	return &ddlPullerImpl{
+		puller: plr,
+		// the puller will listen changed events from `startTs`(including `startTs`)
```
Comment (liuzix) [inline cdc/owner/ddl_puller.go:76]:
```suggestion
		// the puller will listen to change events from `startTs` (including `startTs`)
```

File: cdc/owner/ddl_puller.go [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+		// so `startTs - 1` is a available resolvedTS, it means that all txn before `startTs - 1` is received(or don't need to receive)
```
Comment (liuzix) [inline cdc/owner/ddl_puller.go:77]:
```suggestion
		// `startTs - 1` is a valid resolvedTS, which means that all transactions before `startTs - 1` have been received (or don't need to be received)
```

File: cdc/owner/ddl_puller.go [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+		cancel:     func() {},
+	}, nil
+}
+
+const ddlPullerName = "DDL_PULLER"
+
+func (h *ddlPullerImpl) Run(ctx cdcContext.Context) error {
+	ctx, cancel := cdcContext.WithCancel(ctx)
+	h.cancel = cancel
+	log.Debug("DDL puller started")
```
Comment (liuzix) [inline cdc/owner/ddl_puller.go:89]: Some more information in the log?

File: cdc/owner/ddl_puller.go [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+
+	errg.Go(func() error {
+		return h.puller.Run(ctx)
+	})
+
+	rawDDLCh := puller.SortOutput(ctx, h.puller.Output())
+
+	receiveDDL := func(rawDDL *model.RawKVEntry) error {
+		if rawDDL == nil {
+			return nil
```
Comment (liuzix) [inline cdc/owner/ddl_puller.go:102]: Is this branch ever reachable in production?

File: cdc/owner/ddl_puller.go [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+				h.resolvedTS = rawDDL.CRTs
+			}
+			return nil
+		}
+		job, err := entry.UnmarshalDDL(rawDDL)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if job == nil {
+			return nil
```
Comment (liuzix) [inline cdc/owner/ddl_puller.go:117]: Is this reachable?

File: cdc/owner/ddl_puller.go [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+		}
+		if h.filter.ShouldDiscardDDL(job.Type) {
+			log.Info("discard the ddl job", zap.Int64("jobID", job.ID), zap.String("query", job.Query))
+			return nil
+		}
+		h.mu.Lock()
+		defer h.mu.Unlock()
+		h.pendingDDLJobs = append(h.pendingDDLJobs, job)
```
Comment (liuzix) [inline cdc/owner/ddl_puller.go:125]: Assume a very large amount of DDLs, could appending to the same slice cause memory leak?

File: cdc/owner/ddl_puller.go (hunk 15 above)
Comment (leoppro) [reply cdc/owner/ddl_puller.go:102]: should never reachable, just for safe

File: cdc/owner/ddl_puller.go (hunk 16 above)
Comment (leoppro) [reply cdc/owner/ddl_puller.go:117]: should never reachable, just for safe

File: cdc/owner/ddl_puller.go (hunk 17 above)
Comment (leoppro) [reply cdc/owner/ddl_puller.go:125]: I think it's not a bug problem. the leak is controllable if it is exist

**Pull request merged**